homeassistant/components/recorder.py

```diff
@@ -404,6 +410,29 @@ def save_migration(migration_id):
 
             save_migration(4)
 
+        if migration_id < 5:
+            # Add domain so that thermostat graphs look right
+            self.query("""
+                ALTER TABLE states
+                ADD COLUMN domain text
+            """)
+
+            # populate domain with defaults
+            self.query("""
+                UPDATE states
+                set domain=substr(entity_id, 0, instr(entity_id, '.'))
+            """)
+
+            # add indexes we are going to use a lot on selects
+            self.query(
```

Comment:
Please use multi-column indices to support the queries that we will run on this table.

Old query would target
`last_changed=last_updated AND last_updated > ? AND last_updated < ? AND entity_id = ? `

The new query adds the following where condition:
`domain IN (?) and last_updated > ? AND last_updated < ? AND entity_id = ?`

Usually SQL engines will only ever use 1 index to support their query but it seems that SQLite will issue each part of the query separate and union the results ([source](https://www.sqlite.org/queryplanner.html#fig14)). This means that we need to create an index for each query.

I would suggest the following two indices:
```SQL
CREATE INDEX states__state_changes ON states (last_changed, last_updated, entity_id);
CREATE INDEX states__significant_changes ON states (domain, last_updated, entity_id);
```

After adding these indices, we should verify that they are indeed correctly used by calling `EXPLAIN QUERY PLAN SELECT …` ([docs](https://www.sqlite.org/eqp.html))